server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -1005,19 +1012,37 @@ private void setUpdateTablet(UpdateSession us, KeyExtent keyExtent) {
         updateMetrics.addPermissionErrors(0);
         return;
       }
+
     }
 
     @Override
     public void applyUpdates(TInfo tinfo, long updateID, TKeyExtent tkeyExtent,
-        List<TMutation> tmutations) {
+        List<TMutation> tmutations) throws TException {
       UpdateSession us = (UpdateSession) sessionManager.reserveSession(updateID);
       if (us == null) {
         return;
       }
 
       boolean reserved = true;
+      boolean allowWriteThreadSemaphore = false;
       try {
         KeyExtent keyExtent = new KeyExtent(tkeyExtent);
+        maxThreads = getServerConfig().getConfiguration().getCount(Property.TSERV_MAX_WRITETHREADS);
+        maxThreadPermits = maxThreads == 0 ? Integer.MAX_VALUE : maxThreads;
+        sem = new Semaphore(maxThreadPermits);
```

Comment:
The semaphore should be shared between function calls and thread.  This is continually creating new semaphores, so different threads will have different semaphores. With this pattern the semaphore object aquired from could change before the thread returns.  I I think it should be created once for the tserver.

Reply:
Ok, I can move the semaphore to the constructor. I had done that initially, but I was thinking that we may want to change the maxThreadPermits property value without restarting the tablet server.

Reply:
I moved the semaphore to a synchronized getter method that only allows for one creation in my latest commit. 